src/Builder/FrontendDialectTransformer.cpp

```diff
@@ -533,8 +557,17 @@ class FrontendGenImpl {
       // Create a return operation to return all ONNX output tensors.
       builder_.create<ONNXYieldOp>(UnknownLoc(), retVals);
 
+    SmallVector<llvm::StringRef> inputDimParamsRefs, outputDimParamsRefs;
+    for (uint64_t i = 0; i < inputDimParams.size(); ++i)
+      inputDimParamsRefs.emplace_back(llvm::StringRef(inputDimParams[i]));
+    for (uint64_t i = 0; i < outputDimParams.size(); ++i)
+      outputDimParamsRefs.emplace_back(llvm::StringRef(outputDimParams[i]));
     op->setAttr("input_names", builder_.getStrArrayAttr(inputNames));
     op->setAttr("output_names", builder_.getStrArrayAttr(outputNames));
+    op->setAttr(
+        "input_dim_params", builder_.getStrArrayAttr(inputDimParamsRefs));
+    op->setAttr(
+        "output_dim_params", builder_.getStrArrayAttr(outputDimParamsRefs));
```

Comment:
I think it would be nicer if we made all of these argument and result attributes
https://github.com/sorenlassen/llvm-project/blob/main/mlir/include/mlir/Dialect/Func/IR/FuncOps.td#L241-L245
```
func.func @main_graph(
  %arg0: tensor<?x?xf32> {onnx.name = "X", onnx.dims = "0:batch_size,1:sequence_len"},
  %arg1: tensor<1x?xi64> {onnx.name = "Y", onnx.dims = "1:sequence_len"}
) -> (
  tensor<?x?xf32> : {onnx.name = "Z", onnx.dims = "0:batch_size,1:sequence_len"}
)
```
and then you can omit the onnx.dims attribute in the cases where it's empty

Reply:
Great suggestion! I moved to use argument/result attributes instead of function attributes. Thanks!